src/native/fs/fs_napi.cpp

```diff
@@ -1307,18 +1314,67 @@ struct FileReplacexattr : public FSWrapWorker<FileWrap>
 struct LinkFileAt : public FSWrapWorker<FileWrap>
 {
     std::string _filepath;
-
+    int _replace_fd;
     LinkFileAt(const Napi::CallbackInfo& info)
         : FSWrapWorker<FileWrap>(info)
+        , _replace_fd(0)
     {
         _filepath = info[1].As<Napi::String>();
+        if (info.Length() > 2 && !info[2].IsUndefined()) {
+            _replace_fd = info[2].As<Napi::Number>();
+        }
         Begin(XSTR() << "LinkFileAt " << DVAL(_wrap->_path) << DVAL(_wrap->_fd) << DVAL(_filepath));
     }
     virtual void Work()
     {
         // gpfs_linkat() is the same as Linux linkat() but we need a new function because
         // Linux will fail the linkat() if the file already exist and we want to replace it if it existed.
-        SYSCALL_OR_RETURN(dlsym_gpfs_linkat(_wrap->_fd, "", AT_FDCWD, _filepath.c_str(), AT_EMPTY_PATH));
+        std::string cwd = get_current_dir_name();
+        DBG0("LINK_FILE_AT" << DVAL(cwd) << DVAL(_filepath.c_str()));
+        SYSCALL_OR_RETURN(dlsym_gpfs_linkat(_wrap->_fd, "", AT_FDCWD, _filepath.c_str(), AT_EMPTY_PATH, _replace_fd));
+    }
+};
+
+struct UnlinkFileAt : public FSWrapWorker<FileWrap>
+{
+    std::string _filepath;
+    int _delete_fd;
+    UnlinkFileAt(const Napi::CallbackInfo& info)
+        : FSWrapWorker<FileWrap>(info)
+        , _delete_fd(0)
```

Comment:
how about renaming to replace_fd like it is in libgpfs header?

also can you add a comment to document this fd argument - e.g that it's optional, and if undefined it's like 0 and mean what exactly - that any inode can be replaced? or that no file must be replaced - so it works like posix un/linkat? i am not sure which... and when fd is non 0 it will only verify the replaced inode matches before replacing it, but what about if there is no file there - does it proceed with un/linkat? or returns with error?

Reply:
Sure I can change it but in unlinkat() the verified fd is actually called _fd (the verified fd called _replace_fd only in linkat) so should I use replace_fd term here also?